File: samples/mastra-nextjs-postgres-redis/app/src/worker.ts
import { Job, Worker } from "bullmq";

import { classifyItem, embedTextForSearch, generateSeedItems, textForEmbedding } from "@/lib/ai";
import { ensureSchema } from "@/lib/db";
import {
  failSeedRun,
  finishSeedRun,
  getItemById,
  insertSeedItems,
  markItemProcessed,
  startSeedRun,
  updateProcessedItem,
} from "@/lib/items";
import { QUEUE_NAME, getRedisConnection, getSyncQueue } from "@/lib/queue";

type SeedBatchJob = {
  runId: string;
};

type ClassifyItemJob = {
  runId: string;
  itemId: number;
};

async function handleSeedBatch(job: Job<SeedBatchJob>) {
  const { runId } = job.data;
  await startSeedRun(runId, "Generating 10 tasks and 10 events with the LLM");

  const rawItems = await generateSeedItems();
  const insertedItems = await insertSeedItems(runId, rawItems);
  const queue = getSyncQueue();

  for (const item of insertedItems) {
    await queue.add(
      "classify-item",
      {
        runId,
        itemId: item.id,
      } satisfies ClassifyItemJob,
      {
        jobId: `classify:${runId}:${item.id}`,
        attempts: 2,
        backoff: {
          type: "exponential",
          delay: 1000,
        },
        removeOnComplete: 100,
        removeOnFail: 100,
      },
    );
  }

  await startSeedRun(runId, `Queued ${insertedItems.length} background classification jobs`);

  return {
    ok: true,
    itemCount: insertedItems.length,
  };
}

async function handleClassifyItem(job: Job<ClassifyItemJob>) {
  const { runId, itemId } = job.data;
  const item = await getItemById(itemId);

  if (!item) {
    throw new Error(`Item ${itemId} not found`);
  }

  const classification = await classifyItem(item);
  const embedding = await embedTextForSearch(textForEmbedding(item, classification));
  await updateProcessedItem(item.id, classification, embedding);
  const run = await markItemProcessed(runId);

  if (run && run.processedItems >= run.totalItems) {
    await finishSeedRun(runId, `Generated ${run.totalItems} items and classified every one of them`);
  }

  return {
    ok: true,
    itemId,
  };
}

async function main() {
  await ensureSchema();

  const worker = new Worker(
    QUEUE_NAME,
    async (job) => {
      if (job.name === "seed-batch") {
        return handleSeedBatch(job as Job<SeedBatchJob>);
      }

      if (job.name === "classify-item") {
        return handleClassifyItem(job as Job<ClassifyItemJob>);
      }

      return { ok: false, ignored: true };
    },
    {
      connection: getRedisConnection(),
      concurrency: Number(process.env.WORKER_CONCURRENCY ?? 8),
    },
  );

  worker.on("failed", async (job, error) => {
    const runId = typeof job?.data === "object" && job.data && "runId" in job.data ? String(job.data.runId) : null;
    if (runId) {
      await failSeedRun(runId, error.message);
    }
    console.error(`Worker job failed (${job?.name ?? "unknown"}:${job?.id ?? "n/a"}):`, error.message);
  });

  console.log("tasks-and-events worker is running");
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: samples/pydantic-ai-extended/app/worker.py
"""
Background worker that processes jobs from the Redis queue.

Job types:
  - seed-batch: Generate sample tickets and alerts, insert them, then fan out
    per-item classify jobs.
  - classify-item: Classify a single item and generate its embedding.
"""

from __future__ import annotations

import asyncio
import sys
import traceback

from lib.ai import classify_item, embed_text_for_search, generate_seed_items, text_for_embedding
from lib.db import ensure_schema
from lib.items import (
    ItemClassification,
    fail_seed_run,
    finish_seed_run,
    get_item_by_id,
    insert_seed_items,
    mark_item_processed,
    start_seed_run,
    update_processed_item,
)
from lib.queue import dequeue_job, enqueue_job


async def handle_seed_batch(run_id: str) -> None:
    await start_seed_run(run_id, "Generating 10 tickets and 10 alerts with the LLM")

    raw_items = await generate_seed_items()
    inserted = await insert_seed_items(run_id, raw_items)

    for item in inserted:
        await enqueue_job("classify-item", {"runId": run_id, "itemId": item.id})

    await start_seed_run(
        run_id, f"Queued {len(inserted)} background classification jobs"
    )


async def handle_classify_item(run_id: str, item_id: int) -> None:
    item = await get_item_by_id(item_id)
    if item is None:
        raise ValueError(f"Item {item_id} not found")

    item_dict = {
        "item_type": item.item_type,
        "source": item.source,
        "title": item.title,
        "body": item.body,
    }
    classification = await classify_item(item_dict)
    embedding = await embed_text_for_search(text_for_embedding(item_dict, classification))
    await update_processed_item(item.id, classification, embedding)
    run = await mark_item_processed(run_id)

    if run and run.processed_items >= run.total_items:
        await finish_seed_run(
            run_id,
            f"Generated {run.total_items} items and classified every one of them",
        )


async def main() -> None:
    await ensure_schema()
    print("ticket-sync worker is running", flush=True)

    while True:
        try:
            job = await dequeue_job(timeout=5)
            if job is None:
                continue

            name = job.get("name")
            data = job.get("data", {})

            if name == "seed-batch":
                await handle_seed_batch(data["runId"])
            elif name == "classify-item":
                await handle_classify_item(data["runId"], data["itemId"])
            else:
                print(f"Unknown job type: {name}", flush=True)

        except Exception:
            traceback.print_exc()
            run_id = None
            if isinstance(job, dict):
                run_id = job.get("data", {}).get("runId")
            if run_id:
                try:
                    await fail_seed_run(run_id, traceback.format_exc()[-200:])
                except Exception:
                    pass


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        sys.exit(0)
